randovania/generator/old_generator_reach.py

```diff
@@ -43,27 +43,37 @@ def _extra_requirement_for_node(
     return extra_requirement
 
 
-class GraphPath(NamedTuple):
-    previous_node: GraphOrClassicNode | None
-    node: GraphOrClassicNode
+class GraphPath:
+    __slots__ = ("previous_node", "node", "requirement")
+    previous_node: NodeIndex | None
+    node: NodeIndex
     requirement: Requirement
 
+    def __init__(self, previous: NodeIndex | None, node: NodeIndex, requirement: Requirement):
+        self.previous_node = previous
+        self.node = node
+        self.requirement = requirement
+
     def is_in_graph(self, digraph: graph_module.BaseGraph) -> bool:
         if self.previous_node is None:
             return False
         else:
-            return digraph.has_edge(self.previous_node.node_index, self.node.node_index)
+            return digraph.has_edge(self.previous_node, self.node)
 
     def add_to_graph(self, digraph: graph_module.BaseGraph) -> None:
-        digraph.add_node(self.node.node_index)
+        digraph.add_node(self.node)
         if self.previous_node is not None:
-            digraph.add_edge(self.previous_node.node_index, self.node.node_index, data=self.requirement)
+            digraph.add_edge(self.previous_node, self.node, data=self.requirement)
 
 
-class _SafeNodes(typing.NamedTuple):
+class _SafeNodes:
     as_list: list[int]
     as_set: set[int]
 
+    def __init__(self, component: Collection[NodeIndex]):
```

Comment:
might as well update the types for `as_list` and `as_set` to use node index too. also would it be worth giving this class slots?